In [1]:
import MeCab
import unidic
print(unidic.DICDIR)
MeCab.Tagger("-d" + unidic.DICDIR)

/home/yamada/.local/lib/python3.6/site-packages/unidic/dicdir


<Swig Object of type 'MeCab::Tagger *' at 0x7f64f8eb7f10>

- よくわからないもの

    
    

In [2]:
import json
import os
import pandas as pd
import numpy as np
from pathlib import Path
import collections
from sklearn.model_selection import train_test_split
from sklearn import metrics

import sys
sys.path.append("../")
from datatools.analyzer import *
from utterance.error_tools import *

from datatools.maneger import DataManager
from datatools.preproc import Preprocessor

In [3]:
ngram_path = "../../corpus/google/3-gram/"

In [4]:
key1_set = set( os.listdir(ngram_path) ) 

In [17]:
path = "../eval_labeled/"
datalist = ['DCM', 'DIT', 'IRS']
convs = read_conv(path, datalist)

error = "Grammatical error"
errors = ['Grammatical error', "Uninterpretable"]
errors = ['Semantic error', "Wrong information"]
sys_utt = []
y = []
for conv in convs:
    for ut in conv:
        if ut.is_system() and ut.is_exist_error():
            if not ut.utt[-1] in ["？", "！", "。", "!"]:
                sys_utt.append( clean_text( ut.utt+"。" ))
                # sys_utt.append(ut.utt)
            else:   
                sys_utt.append(ut.utt)
            if ut.is_error_included(errors):
                y.append(1)
            else:
                y.append(0)

In [6]:
print("len of y:{0}, error '{1}' counts:{2}".format(len(y), errors, y.count(1)))

len of y:1386, error '['Grammatical error', 'Uninterpretable']' counts:7


In [7]:
def load_dict(path, key1):
    filename = key_formated(key1)
    with open(path+filename, "r") as f:
        key1_dict = json.load(f) 
    return key1_dict  

In [8]:
def key_formated(key1):
    return "ngram_"+key1+".json"

In [9]:
key2dict = dict()
registerform_dict = dict()
analyzer_unidic=  Analyzer(Tokenizer())

In [10]:
#  を考慮
# 何故か↑が出てくる
def judge_key3(key1_dict, ngram):
    key1 = ngram[0]
    key2 = ngram[1]
    key3 = ngram[2]

    if key2 not in key1_dict[key1]:
        # '"'をkeyにして聞いてみる
        if '"' in key1_dict[key1] and key2 in key1_dict[key1]['"']:
            return True
        if '、' in key1_dict[key1] and key2 in key1_dict[key1]['、']:
            return True
        return False
    
    if key3 not in key1_dict[key1][key2]:
        return False

    return True 

In [11]:

def search_registerforam(key, reading):
    filename = key_formated(key)
    if filename in key1_set:
        return key
    # 登録されていない
    else:
        for other in reading:
            filename = key_formated(key)
            if filename in key1_set:
                registerform_dict[key] = other
                return other
        return key


def sentence_judge(sentence, N=3):
    
    df = analyzer_unidic.analyze_with_dataframe(sentence)
    
    # 登録
    sentence_ = ""
    for txt, read in zip(df.surface, df.reading):
        sentence_ += search_registerforam(txt, read)
    sents = sentence2docs(sentence_, sents_span=True)
    # print(sents)

    # print(tokens)
    ngram_text = []
    ngram_pos = []

    for sent in sents:
        # 各文章 : sent
        df = analyzer_unidic.analyze_with_dataframe("".join([token.orth_ for token in sent]))
        surfaces = []
        poses = []
        readings = []
        for txt, p, read in zip(df.surface, df.part_of_speech, df.reading):
            surfaces.append(txt)
            readings.append(read)
            poses.append(p)

            for i in range(len(surfaces)-N+1):
                # print(L[i:i+N])
                ngram_text.append(surfaces[i:i+N])
                ngram_pos.append(poses[i:i+N])
                # ngram_reading.append(readings[i:i+N])
    
    # print(ngram_text)
    # return
    
    for ngram, pgram in zip(ngram_text, ngram_pos):
        # if "助動詞" in pgram[1] or "助詞" in pgram[1] or "助動詞" in pgram[0] or "助詞" in pgram[0]:
        if "助動詞" in pgram[0] or "助詞" in pgram[0]:
            # print("\tcontext : {0}| ->".format(ngram))

            key1 = ngram[0]
            if key_formated(key1) not in key1_set:
                continue

            if key1 in key2dict:
                key1_dict = key2dict[key1]
            else:
                key1_dict = load_dict(ngram_path, key1)
                key2dict[key1] = key1_dict

            # チェック
            if not judge_key3(key1_dict, ngram):
                print(ngram)
                # reading を調整
                
                # 見つかれば，その語彙を登録
                return False
    return True
            


In [12]:
y_pred_2 = []
for i, utt in enumerate( tqdm( sys_utt ) )  :
    cleaned = clean_text(utt).replace("?", "。").replace("!", "")
    # print(cleaned)
    if sentence_judge(cleaned, N=3):
        y_pred_2.append(0)
    else:
        y_pred_2.append(1)
    # if i==35:
    #     break

  0%|          | 2/1386 [00:14<2:42:10,  7.03s/it]

['の', '逆回り', 'は']


  0%|          | 3/1386 [00:19<2:28:23,  6.44s/it]

['と', '、', '頼り甲斐']


  1%|▏         | 20/1386 [00:32<11:07,  2.05it/s] 

['の', 'ソウル', 'らしい']


  2%|▏         | 31/1386 [00:33<03:14,  6.97it/s]

['を', '受け', 'ましょう']


  4%|▍         | 53/1386 [00:35<02:02, 10.90it/s]

['を', '変え', 'ましょう']


  5%|▍         | 67/1386 [00:39<03:24,  6.46it/s]

['ね', '、', '渾名']
['ね', '、', '渾名']


  6%|▋         | 88/1386 [00:40<01:21, 15.84it/s]

['は', '一人', 'が']
['で', '御洒落', 'し']


  7%|▋         | 96/1386 [00:40<00:57, 22.29it/s]

['も', '頑張ろう', 'って']
['ん', 'でしょう', 'か']


  8%|▊         | 112/1386 [00:40<00:41, 30.58it/s]

['で', '御洒落', 'し']
['で', 'どんな', 'に']
['に', '格好', '付き']


  8%|▊         | 116/1386 [00:41<01:03, 19.87it/s]

['なら', 'を', '使う']


  9%|▉         | 130/1386 [00:41<00:52, 24.08it/s]

['たい', 'ん', 'でしょう']


 10%|▉         | 138/1386 [00:42<00:44, 27.91it/s]

['てる', 'ん', 'でしょう']
['を', '損ない', 'ましょう']


 12%|█▏        | 160/1386 [00:43<01:07, 18.18it/s]

['の', '朝御飯', 'メニュー']


 12%|█▏        | 163/1386 [00:43<01:43, 11.87it/s]

['より', '0', '打']


 13%|█▎        | 176/1386 [00:44<00:54, 22.28it/s]

['じゃあ', '漫画', 'も']
['は', '辰年', 'です']
['です', 'か', '…']


 14%|█▍        | 199/1386 [00:45<00:42, 28.13it/s]

['と', 'ホッと', 'する']


 15%|█▌        | 212/1386 [00:45<00:34, 34.45it/s]

['は', '食', '生活']


 16%|█▌        | 225/1386 [00:46<00:36, 31.50it/s]

['の', '大きい', '福耳']


 18%|█▊        | 246/1386 [00:47<00:39, 28.51it/s]

['の', 'お', '昼']


 18%|█▊        | 255/1386 [00:47<00:33, 33.69it/s]

['で', 'テレビ', 'ある']


 19%|█▉        | 268/1386 [00:47<00:33, 33.36it/s]

['は', '一人', 'で']
['に', '欠かせ', 'て']
['は', '不', '得意']


 22%|██▏       | 310/1386 [00:48<00:25, 42.57it/s]

['た', 'を', '仕入れる']
['た', 'は', '鮮度']


 24%|██▍       | 339/1386 [00:49<00:26, 39.20it/s]

['で', 'iPhone', '0']
['へん', 'せき', 'が']


 25%|██▌       | 347/1386 [00:49<00:30, 34.35it/s]

['て', 'いき', 'ましょう']
['と', 'クリスマス', 'カード']


 27%|██▋       | 377/1386 [00:50<00:40, 24.81it/s]

['で', '御洒落', 'し']


 29%|██▊       | 398/1386 [00:51<00:32, 30.16it/s]

['から', '中', '学校']
['迄', '旅行', 'に']


 29%|██▉       | 402/1386 [00:51<00:36, 27.30it/s]

['は', 'くま', 'もん']
['と', 'いう', 'アイス']
['の', 'ギョーザ', '等']
['の', 'ウッ', 'チャン']


 30%|██▉       | 409/1386 [00:52<00:56, 17.22it/s]

['と', 'ハビエル', '・']
['と', '八木', '竜一']
['た', '映画', 'フォーブス']
['は', '違和', '感']


 30%|██▉       | 415/1386 [00:52<00:50, 19.31it/s]

['た', 'ドトール', 'コーヒー']
['で', 'ジャイアン', '選手']
['で', '初めて', 'エボラ']


 30%|███       | 422/1386 [00:52<00:43, 22.24it/s]

['の', '笑点', 'は']
['の', 'ファミリー', 'マート']
['に', '住む', '大渕']
['は', 'の', 'パンケーキ']
['で', 'カフェ', '坂の下']


 31%|███       | 428/1386 [00:53<00:41, 22.82it/s]

['と', 'いう', '名護']
['な', 'ゴーヤー', '等']


 31%|███▏      | 434/1386 [00:53<00:41, 22.84it/s]

['の', '御', '店']
['の', '御', '店']
['の', '振り掛け', '・']


 32%|███▏      | 440/1386 [00:53<00:39, 23.78it/s]

['ない', '物真似', '選手']
['の', '海遊', '館']
['た', '、', '蛸焼き']
['の', 'で', '海遊']


 32%|███▏      | 446/1386 [00:53<00:39, 24.04it/s]

['から', '、', '海遊']
['の', '高級', 'ポッキー']
['の', '高級', 'ポッキー']
['の', 'チョコレートラテポッキーブレンド', 'が']


 32%|███▏      | 449/1386 [00:54<00:42, 21.92it/s]

['が', '0', '数']
['の', '子', '大粒']
['から', 'でしょう', 'か']


 33%|███▎      | 452/1386 [00:54<00:42, 22.09it/s]

['は', 'コロ', 'プラ']
['の', 'shall', 'ウィー']


 33%|███▎      | 457/1386 [00:55<01:36,  9.62it/s]

['の', 'で', 'ミスター']
['の', 'は', '一握り']
['で', 'の', '饂飩']
['の', 'キャンデー', 'Microsoftband']


 33%|███▎      | 461/1386 [00:55<01:17, 11.88it/s]

['の', 'キャンデー', 'Microsoftband']
['で', 'ある', 'Microsoftband']
['の', '里田', 'まい']
['で', 'ある', 'Microsoftband']


 34%|███▎      | 467/1386 [00:55<00:55, 16.45it/s]

['だ', 'から', '荒野']
['タ', 'モリ', 'さん']
['の', '町歩き', 'テレビ']
['の', '町歩き', 'テレビ']


 34%|███▍      | 474/1386 [00:55<00:46, 19.77it/s]

['は', '、', 'ashtsujiguchi']
['な', 'スイーツ', '・']
['れる', 'ミディアム', '霊能']
['た', 'お', '節']


 35%|███▍      | 480/1386 [00:56<00:39, 22.70it/s]

['が', '、', '糠床']
['は', 'は', 'ハードル']
['の', '御', '馳走']
['の', '町歩き', 'テレビ']
['タ', 'レント', 'の']


 35%|███▍      | 483/1386 [00:56<00:53, 16.87it/s]

['と', 'いう', 'ホール']
['た', 'ワールド', 'カップ']
['た', 'イングレス', 'の']
['と', 'コンビニエンス', 'ストア']
['の', '遠藤', '保仁']


 35%|███▌      | 489/1386 [00:56<00:48, 18.46it/s]

['から', 'はずれ', '淋しく']
['の', '世界', 'ナゼ']
['て', 'も', '仕方ない']
['と', '遠刈田', '温泉']


 36%|███▌      | 495/1386 [00:57<00:41, 21.36it/s]

['タ', 'モリ', 'さん']
['の', '町歩き', 'テレビ']


 36%|███▌      | 501/1386 [00:57<00:52, 16.92it/s]

['の', '八木', '竜']
['の', '中華', 'まん']
['の', 'ハイ', 'チュウ']
['は', 'ヨーク', 'ベニマル']


 37%|███▋      | 507/1386 [00:57<00:42, 20.59it/s]

['は', '表', '参道']
['に', '表', '参道']
['の', '村上', '信五']
['の', '村上', '信五']
['が', '持つ', 'レギュラー']
['は', '、', '牛タン']


 37%|███▋      | 513/1386 [00:57<00:40, 21.61it/s]

['は', '所々', 'たくさん']
['と', '褌姿', 'は']
['と', 'いう', 'ドラマ']
['が', '、', '糠床']


 38%|███▊      | 520/1386 [00:58<00:33, 25.55it/s]

['だ', 'から', '荒野']
['と', 'いう', '旅']
['タ', 'レント', 'の']
['の', 'ブラタモリ', 'で']
['の', '閻', '涵']


 38%|███▊      | 529/1386 [00:58<00:33, 25.80it/s]

['の', 'や', '別ち']
['と', '年末', '年始']
['か', 'も', 'ん']


 39%|███▊      | 535/1386 [00:58<00:32, 26.16it/s]

['か', 'も', 'ん']
['が', '、', '日清']
['の', 'ソレデモシタイ', 'の']
['と', '建て', '前編']


 39%|███▉      | 541/1386 [00:59<00:31, 26.81it/s]

['で', 'は', '日清']
['の', '子', '大粒']
['は', '結構', '太麺']
['も', 'サッカー', 'オーストラリア']
['が', 'シドニー', '球界']


 39%|███▉      | 544/1386 [00:59<00:30, 27.46it/s]

['の', 'シドニー', '球界']
['が', 'シドニー', '代表']
['は', 'サッポロ', 'ビール']


 40%|███▉      | 550/1386 [00:59<00:36, 22.81it/s]

['を', '約', '00000']
['も', 'サッカー', 'オランダ']


 40%|████      | 556/1386 [00:59<00:36, 22.63it/s]

['は', 'ジャニーズ', 'の']
['と', '有名', '人']
['て', 'いる', 'ジャニーズ']
['の', '忘年', '会']
['と', 'ローリー', 'なら']


 41%|████      | 563/1386 [00:59<00:33, 24.36it/s]

['と', 'いう', 'スイーツ']
['の', '御薩', 'スナック']
['の', 'ちょい', '足し']
['に', '印刷', '物']
['と', 'いう', 'スイーツ']


 41%|████      | 569/1386 [01:00<00:33, 24.64it/s]

['の', '0', '週間']
['を', '0', '名']
['が', '和洋', '折衷']
['の', 'かっぱ', '蝦煎']


 42%|████▏     | 579/1386 [01:00<00:30, 26.89it/s]

['が', '楽しい', 'でしょう']
['よう', 'な', 'かば']
['の', 'グランド', 'カルビー']
['の', '力め', 'し']
['と', 'いう', '円柱']


 42%|████▏     | 585/1386 [01:00<00:33, 23.87it/s]

['が', '桜ん坊', 'の']
['の', '0', '県']
['の', '0', '県']
['に', '、', 'こて']


 42%|████▏     | 588/1386 [01:00<00:33, 23.53it/s]

['の', 'で', '滅多']
['の', 'ハイ', 'チュウ']
['の', '広島', '鉄']
['たい', 'の', 'でしょう']
['は', '美味しい', 'ナゴヤ']


 43%|████▎     | 595/1386 [01:01<00:32, 24.71it/s]

['の', '安納', '薯']
['の', '安納', '薯']
['の', 'ファミマ', 'プレミアム']
['の', '、', 'ウォーズ']


 43%|████▎     | 601/1386 [01:01<00:32, 23.85it/s]

['は', 'メイ', 'トー']
['の', '羽生', '結']
['も', 'サッカー', 'オーストラリア']
['で', 'アカデミー', '名誉']


 44%|████▍     | 607/1386 [01:01<00:33, 23.55it/s]

['の', 'サッカー', 'ホンジュラス']
['の', '田中', '将大']
['な', 'サンタ', 'クロース']
['の', 'ミスター', 'クロワッサン']


 44%|████▍     | 613/1386 [01:01<00:30, 25.19it/s]

['の', 'ミスター', 'クロワッサン']
['て', 'くれる', 'でしょう']
['の', '子', '大粒']
['れる', 'ウルトラ', 'ヒーロー']


 45%|████▍     | 619/1386 [01:02<00:29, 25.72it/s]

['の', '馬事', '公園']
['の', 'で', '凄い']


 45%|████▍     | 622/1386 [01:02<00:47, 16.12it/s]

['の', 'フローズン', 'ファンタジー']
['に', '日本', 'エレキテル']
['で', 'ある', 'モンスターズ']
['の', 'ハビエル', '・']


 45%|████▌     | 628/1386 [01:02<00:42, 17.91it/s]

['と', '0', 'FIFA']
['から', 'はずれ', '淋しく']
['の', 'サッカー', '横浜']
['も', '0', '位']
['の', 'じゃがりこ', '味']


 46%|████▌     | 635/1386 [01:03<00:35, 21.16it/s]

['な', '御', '店']
['の', 'グランド', 'カルビー']
['て', 'ハロー', 'キティー']
['と', 'いう', 'テレビ']


 46%|████▌     | 638/1386 [01:03<00:34, 21.60it/s]

['と', 'いう', 'テレビ']
['で', '年末', '年始']
['の', 'ザ', 'manzai']
['と', 'いう', 'ドラマ']
['た', '0', 'g']


 46%|████▋     | 644/1386 [01:03<00:35, 20.84it/s]

['の', '痛快', '明石家']
['と', 'ナインティー', 'ナイン']
['たら', '本当', 'は']
['と', 'マドリード', 'なら']


 47%|████▋     | 650/1386 [01:03<00:34, 21.41it/s]

['から', 'でしょう', 'か']
['や', '田丸', '屋']
['の', 'あまおう', 'の']
['な', '0', '組']


 47%|████▋     | 656/1386 [01:04<00:32, 22.26it/s]

['れる', 'カバーアルバムキャンニュキープ', 'a']
['を', '著名', 'アーチスト']
['な', '0', '組']
['など', 'が', 'You']
['は', '現', '時点']
['て', 'シンガポール', 'プロ']


 48%|████▊     | 662/1386 [01:04<00:30, 23.36it/s]

['が', 'シンガポール', '球界']
['と', 'いう', 'ホテル']
['は', 'とくに', '松坂']
['の', 'カミング', 'アウト']
['の', 'カミング', 'アウト']


 48%|████▊     | 665/1386 [01:04<00:29, 24.10it/s]

['の', '御', '店']
['の', 'カミング', 'アウト']
['の', '漆塗り', 'が']


 48%|████▊     | 671/1386 [01:04<00:30, 23.07it/s]

['は', 'もちろん', '付け麺']
['で', 'は', '鸚哥']
['の', '水族', '館']
['て', 'から', 'ディフェンダー']


 49%|████▉     | 677/1386 [01:04<00:30, 23.20it/s]

['と', 'いう', 'ドイツ']
['の', 'EU', '戦']
['の', 'サッカー', 'コロンビア']


 49%|████▉     | 684/1386 [01:05<00:26, 26.50it/s]

['た', 'かば', '館']
['の', '御', '店']
['よう', 'な', 'かば']
['の', '妖怪', 'ウォッチ']
['の', 'フェルメール', 'さん']


 50%|████▉     | 691/1386 [01:05<00:26, 26.03it/s]

['や', 'ハビエル・アギーレ', 'など']
['の', 'ココ', 'ロココ']
['の', 'ココ', 'ロココ']
['の', '世界', 'ナゼ']


 50%|█████     | 697/1386 [01:05<00:29, 23.28it/s]

['た', 'の', 'プラモデル']
['の', 'BANDAI', 'ホビー']
['の', 'BANDAI', 'ホビー']
['と', 'いう', '炭酸']


 51%|█████     | 703/1386 [01:06<00:28, 23.89it/s]

['の', 'ドーナツ', '0']


 51%|█████     | 709/1386 [01:06<00:33, 20.32it/s]

['と', 'スペイン', 'なら']
['の', '本田', '圭佑']
['の', '本田', '圭佑']
['が', '三都主', 'アレッサンドロ']


 52%|█████▏    | 715/1386 [01:06<00:31, 21.18it/s]

['の', 'リプトン', '・']
['で', '銀', 'メダル']
['まで', 'FIFA', 'ワールド']


 52%|█████▏    | 718/1386 [01:06<00:31, 21.17it/s]

['か', '鈴木', '明子']
['で', 'ある', 'イングレス']
['より', 'もう', '少し']
['に', '島歌', 'と']


 52%|█████▏    | 723/1386 [01:07<00:43, 15.34it/s]

['で', '、', 'グラッチェ']
['は', '映画', '煎']
['は', '、', '数十']


 52%|█████▏    | 727/1386 [01:07<00:49, 13.21it/s]

['ながら', 'みけ', 'らん']
['に', '御嶽', '山']
['て', 'フジ', 'テレビ']


 53%|█████▎    | 733/1386 [01:07<00:36, 18.13it/s]

['で', '0', '月']
['と', 'いう', 'ドラマ']
['ない', '物真似', '選手']
['は', 'なかなか', '漫才']
['か', 'くま', 'もん']


 53%|█████▎    | 739/1386 [01:08<00:33, 19.08it/s]

['の', 'リプトン', '・']
['の', 'ミッドフィルダー', '本田']
['の', 'セレッソ', '大阪']
['に', '遠藤', '保仁']
['の', '長友', '佑都']


 54%|█████▍    | 745/1386 [01:08<00:30, 21.30it/s]

['も', '神', '様']
['を', '蜂蜜', '0']


 54%|█████▍    | 752/1386 [01:08<00:27, 23.46it/s]

['の', 'ヒカ', 'キン']
['の', 'が', 'ツイッター']
['の', 'エイプリル', 'フール']
['と', 'アムステルダム', 'なら']


 55%|█████▍    | 759/1386 [01:08<00:24, 25.78it/s]

['の', 'サザン', 'オール']
['など', '茅ヶ崎', '地方']
['は', 'ワールド', 'カップ']
['も', 'ブラタモリ', '見']


 55%|█████▌    | 765/1386 [01:09<00:23, 26.23it/s]

['の', '町歩き', 'テレビ']
['か', '森田', '芳光']
['て', '致命', '傷']


 56%|█████▌    | 772/1386 [01:09<00:23, 25.81it/s]

['は', '海外', '力士']
['が', 'いい', 'でしょう']
['に', 'よる', '0']
['なら', '燐', '・']


 56%|█████▌    | 778/1386 [01:09<00:23, 25.69it/s]

['も', '燐', '・']
['の', '宮本', '恒靖']


 57%|█████▋    | 784/1386 [01:09<00:25, 24.00it/s]

['の', '宮本', '恒靖']
['の', '川澄', '奈穂美']
['の', 'ゴール', 'キーパー']
['の', 'アンデルソン', '・']
['は', 'ガンバ', '大阪']


 57%|█████▋    | 790/1386 [01:10<00:24, 24.59it/s]

['や', '高級', '焼き肉']
['が', '総額', '000000000']
['の', 'ソレデモシタイ', 'は']
['の', 'バラード', 'SHOW']


 57%|█████▋    | 796/1386 [01:10<00:24, 23.61it/s]

['の', 'フィリピン', 'カレー']
['と', 'いう', '音楽']
['と', 'いう', 'サッカー']
['の', '侭', 'は']


 58%|█████▊    | 802/1386 [01:10<00:23, 24.49it/s]

['は', 'テレビ', '朝日']
['は', '、', 'ナインティー']
['れる', 'レコ', 'チョク']
['の', 'ぶー', '社']
['の', 'ぶー', '社']
['と', '国営', '備北']


 58%|█████▊    | 808/1386 [01:10<00:23, 24.61it/s]

['が', '桜ん坊', 'の']
['くらい', 'フーズ', 'に']
['くらい', 'フーズ', 'の']
['の', 'ソントン', 'ホールディングス']
['は', 'アイドル', 'グループス']


 59%|█████▊    | 814/1386 [01:11<00:22, 25.88it/s]

['も', 'とても', 'ヒット']
['な', '0', '組']
['な', '0', '組']
['と', 'いう', 'CD']


 59%|█████▉    | 820/1386 [01:11<00:24, 22.93it/s]

['に', 'よる', '0']
['を', '著名', 'アーチスト']
['と', 'いう', 'CD']
['で', '国営', '滝野']
['から', '夜更かし', 'で']


 59%|█████▉    | 823/1386 [01:11<00:25, 22.17it/s]

['を', 'マツコ', 'の']
['と', 'いう', '技']
['の', '鸚哥', '味']
['の', 'グランド', 'カルビー']


 60%|█████▉    | 830/1386 [01:11<00:21, 25.39it/s]

['の', 'ぐり', 'こ']
['が', '少な', '目']
['の', 'ジャイアントカプリコ', '苺味']
['の', 'ジャイアントカプリコ', '苺味']
['の', '広島', '鉄']


 61%|██████    | 839/1386 [01:12<00:22, 24.57it/s]

['なら', '、', '生チョコ']
['に', 'ポテト', 'チップス']
['の', 'スナック', '菓子']
['で', 'アサヒ', 'ビール']
['の', 'スナック', '菓子']


 61%|██████    | 842/1386 [01:12<00:21, 25.13it/s]

['の', 'かっぱ', '蝦煎']
['の', '三都主', 'アレッサンドロ']
['た', '国際', 'アンデルセン']


 61%|██████    | 848/1386 [01:12<00:30, 17.79it/s]

['と', 'いう', 'ポップコーン']


 62%|██████▏   | 854/1386 [01:12<00:24, 21.42it/s]

['や', '香川', '真司']
['て', 'カミング', 'アウト']
['も', '世界', '選手']
['の', '漆塗り', 'が']


 62%|██████▏   | 860/1386 [01:13<00:22, 23.64it/s]

['の', 'ファミリー', 'マート']
['な', 'コンビニエンス', 'ストア']
['の', 'fukase', 'さん']


 63%|██████▎   | 867/1386 [01:13<00:18, 28.19it/s]

['は', 'まだ', 'エボラ']
['に', '0', '週間']
['ば', '、', 'letao']


 63%|██████▎   | 873/1386 [01:13<00:20, 25.19it/s]

['で', '下', '半身']
['の', '0', '種類']


 63%|██████▎   | 879/1386 [01:13<00:20, 24.84it/s]

['が', '桜ん坊', 'の']
['し', '、', 'グラッチェ']
['だ', 'から', '荒野']


 64%|██████▍   | 885/1386 [01:14<00:19, 25.84it/s]

['と', 'ヨーロッパ', 'なら']
['が', '鈴木', '章']
['の', 'もやもや', 'さまぁ']


 65%|██████▍   | 895/1386 [01:14<00:19, 24.72it/s]

['に', '、', 'サザンアイルズ']
['から', '、', 'サザンアイルズ']
['乃', '逢', 'の']
['は', '特に', 'pasco']
['が', 'pasco', 'の']


 65%|██████▍   | 899/1386 [01:14<00:17, 27.34it/s]

['で', 'は', 'pasco']
['の', '信長', '協奏']


 65%|██████▌   | 907/1386 [01:14<00:16, 28.57it/s]

['まで', 'なら', 'アナウンサー']
['の', '鸚哥', '味']


 66%|██████▌   | 913/1386 [01:15<00:17, 26.49it/s]

['の', '鸚哥', '味']
['の', '鸚哥', '味']
['の', '鸚哥', '味']


 66%|██████▋   | 919/1386 [01:15<00:18, 25.14it/s]

['の', '生姜', 'チャイ']
['の', '生姜', 'チャイ']
['と', '愛媛', '地方']
['て', 'いる', '町作り']
['で', 'モ', 'スライス']


 67%|██████▋   | 925/1386 [01:15<00:19, 23.41it/s]

['も', 'MOS', 'バーガー']
['の', 'フローズン', 'ファンタジー']
['で', '恋', 'する']


 67%|██████▋   | 931/1386 [01:15<00:18, 25.04it/s]

['の', 'akb', '0']
['の', '軍師', '官兵衛']
['の', '軍師', '官兵衛']
['で', '、', 'akb']
['な', '岡崎', '慎司']
['で', 'くつろぐ', '画像']


 68%|██████▊   | 938/1386 [01:16<00:15, 28.59it/s]

['が', '岡崎', '慎司']
['の', 'で', '厄介']


 69%|██████▊   | 950/1386 [01:16<00:16, 27.11it/s]

['で', 'ぎりぎり', '何']
['に', 'ある', '御洒落']
['て', '、', '香り付け']
['と', 'か', 'スープ']


 69%|██████▉   | 963/1386 [01:17<00:14, 29.52it/s]

['か', 'も', '必須']
['です', 'ね', '一人']
['と', 'か', '天国']


 70%|███████   | 971/1386 [01:17<00:13, 29.87it/s]

['か', '.', '牛タン']
['や', '、', '学祭']


 71%|███████   | 979/1386 [01:17<00:13, 31.22it/s]

['ね', 'ー', 'ソフト']
['ながら', 'お', 'しまい']
['を', 'し', 'ましょう']


 71%|███████   | 987/1386 [01:17<00:12, 33.08it/s]

['と', 'か', 'いる']
['に', '読む', '積もり']


 72%|███████▏  | 1000/1386 [01:18<00:10, 36.96it/s]

['と', 'か', '操作']
['か', '…', 'どじ']
['ず', 'タイピング', 'し']


 73%|███████▎  | 1009/1386 [01:18<00:10, 34.35it/s]

['と', '柴', '0']
['と', '柴', '0']


 74%|███████▎  | 1021/1386 [01:18<00:10, 34.46it/s]

['か', '…', 'どじ']


 74%|███████▍  | 1025/1386 [01:18<00:10, 34.75it/s]

['ます', 'そう', 'いう']
['てる', 'ん', 'でしょう']
['が', '、', '何に']


 75%|███████▍  | 1034/1386 [01:19<00:10, 33.71it/s]

['です', 'か', '大晦日']
['の', '0', '週間']


 75%|███████▌  | 1042/1386 [01:19<00:11, 30.58it/s]

['も', '0', '週間']
['す', '…', '凄い']


 76%|███████▌  | 1054/1386 [01:19<00:10, 31.03it/s]

['は', 'やみつき', 'に']


 76%|███████▋  | 1058/1386 [01:20<00:13, 24.88it/s]

['た', 'くらい', 'でしょう']
['を', '食べ', 'ましょう']


 77%|███████▋  | 1065/1386 [01:20<00:11, 27.50it/s]

['た', '「', 'あずみ']
['です', 'かー', '。']


 78%|███████▊  | 1082/1386 [01:20<00:08, 34.36it/s]

['の', 'ジャズ', 'セクション']
['と', 'か', '自然']
['の', 'クライン', 'キー']
['に', 'なろう', 'と']
['と', 'か', '自然']


 79%|███████▉  | 1094/1386 [01:21<00:08, 32.69it/s]

['って', '陰', 'から']
['でし', 'た', '…']
['か', 'も', '必須']
['か', 'が', 'ひどかっ']


 80%|███████▉  | 1102/1386 [01:21<00:08, 34.09it/s]

['は', '酎ハイ', '程度']
['は', '酎ハイ', '程度']
['じゃあ', 'あまり', 'お']
['よ', 'ね', '…']


 80%|████████  | 1110/1386 [01:21<00:08, 31.66it/s]

['と', 'か', '表情']
['が', '切磋', '琢磨']


 80%|████████  | 1115/1386 [01:21<00:07, 35.42it/s]

['に', '読む', '積もり']
['の', '踏み場', 'は']


 82%|████████▏ | 1131/1386 [01:22<00:07, 32.04it/s]

['が', 'やたら', '粘る']
['で', '忘年', '会']
['です', '.', '。']


 82%|████████▏ | 1135/1386 [01:22<00:08, 30.93it/s]

['と', 'いう', '斬新']
['が', '声高い', '子']


 83%|████████▎ | 1148/1386 [01:22<00:06, 35.85it/s]

['に', '瀕し', 'て']
['てる', 'ん', 'でしょう']
['な', 'ん', 'でしょう']


 84%|████████▎ | 1160/1386 [01:23<00:06, 35.55it/s]

['です', 'ね', '…']
['と', 'か', '天国']


 85%|████████▍ | 1172/1386 [01:23<00:05, 36.38it/s]

['が', '若槻', 'っぽい']
['ない', 'ん', 'でしょう']


 85%|████████▌ | 1180/1386 [01:23<00:05, 35.53it/s]

['て', '焦がし', 'たい']
['の', '白土', '人']
['から', 'そう', '急かす']


 87%|████████▋ | 1201/1386 [01:24<00:04, 37.22it/s]

['たら', '猫好き', 'に']


 87%|████████▋ | 1209/1386 [01:24<00:04, 36.03it/s]

['だ', 'か', '取り留め']
['で', '済ま', 'せよう']
['っけ', '、', '時代']
['です', 'けど', '名店']


 88%|████████▊ | 1214/1386 [01:24<00:04, 38.04it/s]

['なんか', 'ダイソン', 'です']


 89%|████████▉ | 1235/1386 [01:25<00:04, 33.39it/s]

['は', '痛み', 'にく']
['で', 'たまに', 'マージャン']


 90%|████████▉ | 1243/1386 [01:25<00:04, 31.97it/s]

['れる', 'こと', 'でしょう']
['よっ', '笑い', '。']


 90%|█████████ | 1252/1386 [01:25<00:03, 33.62it/s]

['も', 'オリンピック', '書い']
['す', 'っ', 'っ']
['に', '一人', 'だけ']
['と', '一人', 'で']


 91%|█████████ | 1260/1386 [01:26<00:03, 32.67it/s]

['と', 'か', '外']
['て', '食べれ', 'ます']


 91%|█████████▏| 1268/1386 [01:26<00:03, 31.70it/s]

['の', 'で', '短所']
['な', 'マスコット', 'キャラクター']


 93%|█████████▎| 1284/1386 [01:26<00:02, 34.55it/s]

['を', '0', 'くらい']
['も', 'さっさと', '美味しい']
['に', 'は', '敵い']
['です', 'かー', '。']
['です', 'か', '年始']


 93%|█████████▎| 1292/1386 [01:27<00:02, 33.80it/s]

['から', '段', 'ボール']
['ほど', 'あり', '一気']
['か', 'も', '必須']


 94%|█████████▍| 1300/1386 [01:27<00:02, 34.33it/s]

['と', 'か', '日本']
['と', 'か', '絶対']


 94%|█████████▍| 1308/1386 [01:27<00:02, 33.00it/s]

['そう', 'す', 'ね']
['へ', '通っ', 'てらっしゃる']
['と', 'いう', 'オーストラリア']
['が', '切磋', '琢磨']
['ほど', '0', '年']


 95%|█████████▌| 1317/1386 [01:27<00:02, 33.68it/s]

['な', 'ん', 'でしょう']
['が', '切磋', '琢磨']
['みたい', 'に', 'かなう']
['た', 'から', '…']


 96%|█████████▌| 1325/1386 [01:28<00:01, 33.56it/s]

['の', 'で', 'ガッツリ']
['と', '黒猫', 'です']


 96%|█████████▌| 1333/1386 [01:28<00:01, 34.81it/s]

['も', 'ここ', '0']
['は', '大変', 'でしょう']


 97%|█████████▋| 1345/1386 [01:28<00:01, 33.91it/s]

['みたい', 'に', 'かなう']
['で', '…', '。']
['か', 'は', '作り置き']


 98%|█████████▊| 1358/1386 [01:29<00:00, 32.56it/s]

['なら', '"', '良い']
['でし', 'た', '…']
['た', 'くらい', 'でしょう']


 99%|█████████▊| 1366/1386 [01:29<00:00, 32.53it/s]

['が', '勉強', '学校']
['なんて', '個人', 'も']
['と', 'か', '暇']
['の', '踏み場', 'は']


 99%|█████████▉| 1375/1386 [01:29<00:00, 35.23it/s]

['の', '踏み場', 'は']
['は', '仕方ない', 'です']


100%|██████████| 1386/1386 [01:29<00:00, 15.42it/s]

['た', '飛行', '機']


In [13]:
y_pred_2.count(0)

876

In [18]:
score(y, y_pred_2)

confusion matrix = 
 [[758 225]
 [118 285]]
accuracy =  0.7525252525252525
precision =  0.5588235294117647
recall =  0.707196029776675
f1 score =  0.6243154435925521


- 使い方が難しい

        confusion matrix = 
        [[666 713]
        [  6   1]]
        accuracy =  0.48124098124098125
        precision =  0.0014005602240896359
        recall =  0.14285714285714285
        f1 score =  0.0027739251040221915

- ちょっと改善(プログラムミスを修正)

        confusion matrix = 
        [[870 509]
        [  6   1]]
        accuracy =  0.6284271284271284
        precision =  0.00196078431372549
        recall =  0.14285714285714285
        f1 score =  0.0038684719535783366

    - 意外と普通っぽい ngram がない
        - ["を", "覚える", *]

    - スコア化はせず， ngram が存在しなければアウトと仮定した
        - 先行研究のモデルで，そういったボーダーは存在しない

- 用法エラー + 誤情報

        confusion matrix = 
        [[758 225]
        [118 285]]
        accuracy =  0.7525252525252525
        precision =  0.5588235294117647
        recall =  0.707196029776675
        f1 score =  0.6243154435925521
    

    

In [19]:
data_path = "../X_y_data/base_y_pred/"
# data_name = "context_content.pickle"
data_name = "utt_content.pickle"
dataM = DataManager(data_path)
dataM.save_data(data_name, [y, y_pred_2])

success save : ../X_y_data/base_y_pred/utt_content.pickle


In [16]:
df = analyzer_unidic.analyze_with_dataframe("頼りがいのあった可愛くない兄さんが死んだ")
df

surface part_of_speech infl_type infl_form base_form  \
0     頼りがい    名詞-普通名詞-一般-                         タヨリガイ   
1        の       助詞-格助詞--                             ノ   
2       あっ     動詞-非自立可能--     五段-ラ行   連用形-促音便        アル   
3        た         助動詞---     助動詞-タ    連体形-一般         タ   
4      可愛く       形容詞-一般--       形容詞    連用形-一般      カワイイ   
5       ない    形容詞-非自立可能--       形容詞    連体形-一般        ナイ   
6        兄    名詞-普通名詞-一般-                            アニ   
7       さん    接尾辞-名詞的-一般-                            サン   
8        が       助詞-格助詞--                             ガ   
9       死ん        動詞-一般--     五段-ナ行   連用形-撥音便        シヌ   
10       だ         助動詞---     助動詞-タ    連体形-一般         タ   

                             reading phonetic  
0   [頼り甲斐, 頼りがい, タヨリガイ, 頼りがい, タヨリガイ]        *  
1                    [の, の, ノ, の, ノ]        *  
2               [有る, あっ, アッ, ある, アル]        *  
3                    [た, た, タ, た, タ]        *  
4        [可愛い, 可愛く, カワイク, 可愛い, カワイー]        *  
5               [無い, ない, ナイ, ない, ナイ]        *  
6                  [兄, 兄, アニ, 兄, アニ]        *  
7               [さん, さん, サン, さん, サン]        *  
8                    [が, が, ガ, が, ガ]        *  
9               [死ぬ, 死ん, シン, 死ぬ, シヌ]        *  
10                   [た, だ, ダ, だ, ダ]        *